### ISSUE TITLE
How to define random functions on your own types need to be better documented.

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/julia#25605]
and have better error messages.

On 0.6:

```jl
julia> struct F
           v::Float64
       end

julia> rand(F)
ERROR: MethodError: no method matching rand(::MersenneTwister, ::Type{F})
Closest candidates are:
  rand(::AbstractRNG, ::Type, ::Tuple{Vararg{Int64,N}} where N) at random.jl:390
  rand(::AbstractRNG, ::Type, ::Integer, ::Integer...) at random.jl:391
  rand(::MersenneTwister, ::Type{I<:Base.Random.FloatInterval}) where I<:Base.Random.FloatInterval at random.jl:143
  ...
Stacktrace:
 [1] rand(::Type{T} where T) at ./random.jl:283

julia> Base.rand(::MersenneTwister, ::Type{F}) = F(rand())

julia> rand(F)
F(0.5219484358707998)

julia> rand(F, 2,2)
2×2 Array{F,2}:
 F(0.0131819)  F(0.0734573)
 F(0.440896)   F(0.735767)
```

The error message says exactly what is happening, I define the function, everything works.

On 0.7:

```jl
julia> struct F
           v::Float64
       end

julia> rand(F)
ERROR: ArgumentError: Sampler for this object is not defined
Stacktrace:
 [1] Type at ./random/random.jl:106 [inlined]
 [2] rand at ./random/random.jl:193 [inlined]
 [3] rand at ./random/random.jl:194 [inlined]
 [4] rand(::Type{F}) at ./random/random.jl:198
 [5] top-level scope
```

Ok... what do I do... Looking at docs for `rand`, says nothing about `Sampler`. Looking at NEWS.md, nothing about `Sampler`. Looking at docs, nothing about `Sampler`. Looking at `>help? Sampler`,  `No documentation found.`. Let's try the same as in 0.6:

```jl
julia> Base.rand(::MersenneTwister, ::Type{F}) = F(rand())

julia> rand(F)
F(0.15546036169717214)
```

Woho, it works... or?

```jl
julia> rand(F, 2, 2)
ERROR: ArgumentError: Sampler for this object is not defined
Stacktrace:
 [1] Type at ./random/random.jl:106 [inlined]
 [2] rand at ./random/random.jl:193 [inlined]
 [3] rand! at ./random/random.jl:210 [inlined]
 [4] rand! at ./random/random.jl:206 [inlined]
```
Gives up.

**Comment by JeffBezanson** [comment]
This is a fine example of a general theme I've seen a few times. Sometimes people feel a MethodError is too "low level" and user-unfriendly, and that a more application-specific error message should be substituted for it. But I generally think that it's good to know how the system works, in which case a MethodError is actually quite informative and more useful than an ad-hoc message: the formatting is consistent so you get better at reading MethodErrors over time, and more effort has been put in to MethodErrors to make them include more information.

**Comment by KristofferC** [comment]
MbedTLS fails with this error message now: https://travis-ci.org/JuliaWeb/GitHub.jl/jobs/330410030#L1289.

Still no idea how to fix. Was this supposed to be just an optimization or an API change?

**Comment by stevengj** [comment]
I agree with @KristofferC that the issue here is not `MethodError` at all.  The issue is:

* `Sampler` etc should be documented if is needed to extend `rand`
* The problem here is that the `ArgumentError` (not a `MethodError`!) is obscure and needs to be rewritten/replaced.

**Comment by rfourquet** [comment]
I will address the issue.

**Comment by rfourquet** [comment]
> Sometimes people feel a MethodError is too "low level" and user-unfriendly, and that a more application-specific error message should be substituted for it

This is not what happened here, cf. https://github.com/JuliaLang/julia/pull/23964#issuecomment-358921875

**Comment by KristofferC** [comment]
Bump, what is the status on this? I would like to reiterate how awesome it would be if this could be made non-breaking.

**Comment by fredrikekre** [comment]
Bump; documentation for this is *very* much needed now that we start upgrading to v0. 7.

**Comment by rfourquet** [comment]
> Bump; documentation for this is very much needed now that we start upgrading to v0. 7.

Message received!

**Comment by rfourquet** [comment]
> I would like to reiterate how awesome it would be if this could be made non-breaking.

I believe this is not possible while keeping the benefits of the new framework. I would like to reiterate how limited the previous API was; I know you are not a big fan of this change, but hope you will come to appreciate the new possibilities.

**Comment by zsunberg** [comment]
It's still kind of difficult to find and understand the documentation for this. If I google "ERROR: ArgumentError: Sampler for this object is not defined", it brings me here.

When I finally found the right part of the docs (https://docs.julialang.org/en/v1/stdlib/Random/#Generating-values-from-a-collection-1), I was very surprised to find that, instead of learning how to implement a `Sampler`, I needed to implement rand `Random.rand(rng::AbstractRNG, st::Random.SamplerTrivial{MyDist})`. I think the "Sampler for this object is not defined" error message could be greatly improved to point users more quickly toward a way to fix the problem :)
